### PR TITLE
DAT-37: BE - Clean API

### PR DESCRIPTION
**Pull request by thanhphuchuynh**
fix respond err when token not valid
image

###  validate:
- register:
 	name: required,  max=60
	email:required, email
	password: required, gta: 8
image

- login: 
	email:required, email
	password: required, gta: 8
image

- update:
	_id: required
	name: required, max=60
	email: email
	birthday: required
	relationship: max=60
	looking_for: max=60
	media: [string]
	gender: required, max=20
	sex: max=60
	country: required, max=20,
	hobby: [string]
	about: max=256
	like_id: []
	match_id: []
image

### CODE REVIEW

**Comment by minhnhat97kg** [comment]
LGTM

File: internal/app/api/handler/user/user.go [hunk 1]
```diff
@@ -47,7 +52,12 @@ func (h *Handler) SignUp(w http.ResponseWriter, r *http.Request) {
 	err := json.NewDecoder(r.Body).Decode(&userSignup)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go]: Should change if err: =... ; err != nil {}

File: internal/app/api/handler/user/user.go [hunk 2]
```diff
@@ -47,7 +52,12 @@ func (h *Handler) SignUp(w http.ResponseWriter, r *http.Request) {
...
 
 	if err != nil {
-		respond.JSON(w, http.StatusInternalServerError, h.em.InvalidValue.Request)
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.ValidationFailed)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:54]: Log an error before response.
The Request error is okay, use can change the message if necessary

File: internal/app/api/handler/user/user.go [hunk 3]
```diff
@@ -47,7 +52,12 @@ func (h *Handler) SignUp(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if err := validate.Struct(userSignup); err != nil {
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.ValidationFailed)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:60]: Log error before response

File: internal/app/api/handler/user/user.go [hunk 4]
```diff
@@ -63,17 +73,24 @@ func (h *Handler) SignUp(w http.ResponseWriter, r *http.Request) {
...
 // Post handler  login HTTP request
 func (h *Handler) Login(w http.ResponseWriter, r *http.Request) {
+
 	var UserLogin types.UserLogin
+
 	err := json.NewDecoder(r.Body).Decode(&UserLogin)
 
 	if err != nil {
-		respond.JSON(w, http.StatusUnauthorized, h.em.InvalidValue.IncorrectPasswordEmail)
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.ValidationFailed)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:79]: use if err:= ...; err {} incase no new variable is created
Log error before response
should use another error for case parsing request body failed (Ex; the Request error)

File: internal/app/api/handler/user/user.go [hunk 5]
```diff
@@ -63,17 +73,24 @@ func (h *Handler) SignUp(w http.ResponseWriter, r *http.Request) {
...
+
+	if err := validate.Struct(UserLogin); err != nil {
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.ValidationFailed)
 		return
 	}
 
 	user, err := h.srv.Login(r.Context(), UserLogin)
 	if err != nil {
-		respond.JSON(w, http.StatusUnauthorized, h.em.InvalidValue.IncorrectPasswordEmail)
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.IncorrectPasswordEmail)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:91]: use if err:= ...; err {} incase no new variable is created
Log error before response

File: internal/app/api/handler/user/user.go [hunk 6]
```diff
@@ -84,7 +101,7 @@ func (h *Handler) GetUserByID(w http.ResponseWriter, r *http.Request) {
 
 	user, err := h.srv.FindUserById(r.Context(), mux.Vars(r)["id"])
 	if err != nil {
-		respond.JSON(w, http.StatusUnauthorized, h.em.InvalidValue.Request)
+		respond.JSON(w, http.StatusInternalServerError, h.em.InvalidValue.Request)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go]: use if err:= ...; err {} incase no new variable is created
Log error before response

File: internal/app/api/handler/user/user.go [hunk 7]
```diff
@@ -98,13 +115,17 @@ func (h *Handler) UpdateUserByID(w http.ResponseWriter, r *http.Request) {
 	err := json.NewDecoder(r.Body).Decode(&user)
 
 	if err != nil {
-		respond.JSON(w, http.StatusInternalServerError, h.em.InvalidValue.Request)
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.Request)
+		return
+	}
+	if err := validate.Struct(user); err != nil {
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.ValidationFailed)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:123]: Log error before response

File: internal/app/api/handler/user/user.go [hunk 8]
```diff
@@ -98,13 +115,17 @@ func (h *Handler) UpdateUserByID(w http.ResponseWriter, r *http.Request) {
 	err := json.NewDecoder(r.Body).Decode(&user)
 
 	if err != nil {
-		respond.JSON(w, http.StatusInternalServerError, h.em.InvalidValue.Request)
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.Request)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:117]: use if err:= ...; err {} incase no new variable is created
Log error before response

File: internal/app/api/handler/user/user.go [hunk 9]
```diff
@@ -98,13 +115,17 @@ func (h *Handler) UpdateUserByID(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	if err := validate.Struct(user); err != nil {
+		respond.JSON(w, http.StatusBadRequest, h.em.InvalidValue.ValidationFailed)
 		return
 	}
 
 	error := h.srv.UpdateUserByID(r.Context(), user)
 	if error != nil {
-		respond.JSON(w, http.StatusUnauthorized, h.em.InvalidValue.Request)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:107]: use if err:= ...; err {} incase no new variable is created
Log error before response

File: internal/app/api/handler/user/user.go [hunk 10]
```diff
@@ -119,7 +140,7 @@ func (h *Handler) GetListUsers(w http.ResponseWriter, r *http.Request) {
 
 	userList, err := h.srv.GetListUsers(r.Context(), pageParameter, sizeParameter)
 	if err != nil {
-		respond.JSON(w, http.StatusUnauthorized, h.em.InvalidValue.Request)
+		respond.JSON(w, http.StatusInternalServerError, h.em.InvalidValue.Request)
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go:143]: use if err:= ...; err {} incase no new variable is created
Log error before response

File: internal/app/api/types/user.go [hunk 11]
```diff
@@ -7,17 +7,17 @@ import (
 )
 
 type User struct {
-	ID           bson.ObjectId   `json:"_id" bson:"_id,omitempty"`
-	Name         string          `json:"name" bson:"name"`
+	ID           bson.ObjectId   `json:"_id" bson:"_id,omitempty" validate:"required"`
+	Name         string          `json:"name" bson:"name" validate:"required"`
 	Email        string          `json:"email" bson:"email"`
```
Comment (paphuc) [inline internal/app/api/types/user.go]: Should be required this field? Please add Regrex to verify email format

File: internal/app/api/types/user.go [hunk 12]
```diff
@@ -7,17 +7,17 @@ import (
...
-	Age          int             `json:"age" bson:"age"`
-	Relationship string          `json:"relationship" bson:"relationship"`
+	Age          int             `json:"age" bson:"age" validate:"required"`
+	Relationship string          `json:"relationship" bson:"relationship" `
 	LookingFor   string          `json:"lookingFor" bson:"lookingFor"`
```
Comment (paphuc) [inline internal/app/api/types/user.go]: json should be looking_for

File: internal/app/api/types/user.go [hunk 13]
```diff
@@ -7,17 +7,17 @@ import (
...
 	Password     string          `json:"password" bson:"password"`
```
Comment (paphuc) [inline internal/app/api/types/user.go:16]: Should be required this field?

**Comment by paphuc** [comment]
Please add length check for string fiellds

File: internal/app/api/types/user.go [hunk 14]
```diff
@@ -7,17 +7,17 @@ import (
...
 
 type User struct {
-	ID           bson.ObjectId   `json:"_id" bson:"_id,omitempty"`
-	Name         string          `json:"name" bson:"name"`
+	ID           bson.ObjectId   `json:"_id" bson:"_id,omitempty" validate:"required"`
+	Name         string          `json:"name" bson:"name" validate:"required"`
 	Email        string          `json:"email" bson:"email"`
-	Age          int             `json:"age" bson:"age"`
-	Relationship string          `json:"relationship" bson:"relationship"`
+	Age          int             `json:"age" bson:"age" validate:"required"`
```
Comment (paphuc) [inline internal/app/api/types/user.go]: Should change to Birthday with type time

File: internal/app/api/types/user.go (hunk 13 above)
Comment (thanhphuchuynh) [reply internal/app/api/types/user.go:16]: this too used for update, so shouldn't

File: internal/app/api/api.go [hunk 15]
```diff
@@ -60,12 +59,6 @@ func Init(conns *config.Configs, em config.ErrorMessage) (http.Handler, error) {
 	userHandler := userhandler.New(conns, &em, userSrv, userLogger)
 
 	routes := []route{
-		// infra
-		route{
```
Comment (paphuc) [inline internal/app/api/api.go]: should not be removed. We need this API to check our service is still alive

**Pull request merged**